### PR TITLE
[ResourceDetectors.Azure] AppService resource detector

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #.

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        try
+        {
+            serviceName = Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME");
+        }
+        catch
+        {
+            // TODO: log exception.
+        }
+
+        List<KeyValuePair<string, object>>? attributeList = null;
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:42]: Should we consider TagList here?

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/OpenTelemetry.ResourceDetectors.Azure.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>net7.0;net6.0;</TargetFrameworks>
+    <IncludeSharedTestSource>true</IncludeSharedTestSource>
```
Comment (rajkumar-rangaraj) [inline test/OpenTelemetry.ResourceDetectors.Azure.Tests/OpenTelemetry.ResourceDetectors.Azure.Tests.csproj:6]: Just for my understanding, why the property `IncludeSharedTestSource` is used?

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// <copyright file="AzureResourceDetectorTests.cs" company="OpenTelemetry Authors">
```
Comment (rajkumar-rangaraj) [inline test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs:1]: Should we consider naming as AppServiceDetectorTests?

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
+// <copyright file="AppServiceResourceDetector.cs" company="OpenTelemetry Authors">
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:1]: Other projects in this contrib repo has placed this files in Resources folder. Is there a convention followed here?

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+using System;
+using System.Collections.Generic;
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+using Xunit;
+
+namespace OpenTelemetry.ResourceDetectors.Azure.Tests;
+
+public class AzureResourceDetectorTests
+{
```
Comment (rajkumar-rangaraj) [inline test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs:26]: Add a test to with detector and no environment variable.

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            attributeList.Add(new KeyValuePair<string, object>(ResourceSemanticConventions.AttributeServiceName, serviceName));
+
+            // TODO: Add other attributes e.g. service.instance.id
+        }
+        else
+        {
+            return null;
+        }
+
+        return new Resource(attributeList);
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:57]: nit: For reading the code better, we could move the `attributeList` creation and `return new Resource(attributeList);` inside if loop.  We could avoid else and just return null.

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs (hunk 4 above)
Comment (Kielek) [reply src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:1]: I think that it is leftover, from old package name, I would expect have it directly in project folder in this case.

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs (hunk 6 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:57]: Ignore this comment, later realized we will have else if here.

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs (hunk 3 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs:1]: I was not anticipating lot of tests for each detector so, was just planning to include in single file. I can separate it out.

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            attributeList = new();
+
+            attributeList.Add(new KeyValuePair<string, object>(ResourceSemanticConventions.AttributeServiceName, serviceName));
+
+            // TODO: Add other attributes e.g. service.instance.id
+        }
+        else
+        {
+            return null;
```
Comment (Kielek) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:54]: Is it possible make the method returning not null, and here just Resource.Empty?

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:42]: no need. this is one time, and not hot path.

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    [Fact]
+    public void AppServiceResourceDetectorTest()
+    {
+        Environment.SetEnvironmentVariable("WEBSITE_SITE_NAME", "AzureAppService");
+        var resource = ResourceBuilder.CreateEmpty().AddDetector(new AppServiceResourceDetector()).Build();
+        Assert.NotNull(resource);
+        Assert.Contains(new KeyValuePair<string, object>(ResourceSemanticConventions.AttributeServiceName, "AzureAppService"), resource.Attributes);
+        Environment.SetEnvironmentVariable("WEBSITE_SITE_NAME", null);
```
Comment (cijothomas) [inline test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs:34]: finally block

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+/// </summary>
+public sealed class AppServiceResourceDetector : IResourceDetector
+{
+    /// <inheritdoc/>
+    public Resource? Detect()
+    {
+        string? serviceName = null;
+        try
+        {
+            serviceName = Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME");
```
Comment (cijothomas) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:35]: point to the official document which documents this env variable.

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+using System.Collections.Generic;
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.ResourceDetectors.Azure;
+
+/// <summary>
+/// Resource detector for Azure AppService environment.
+/// </summary>
+public sealed class AppServiceResourceDetector : IResourceDetector
```
Comment (cijothomas) [inline src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:27]: Separate PR: See if it is easier to keep this internal, and provide extension methods like on ProviderBuilders, to do `AddAzureResourceDetectors()` instead.

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/OpenTelemetry.ResourceDetectors.Azure.Tests.csproj (hunk 2 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.ResourceDetectors.Azure.Tests/OpenTelemetry.ResourceDetectors.Azure.Tests.csproj:6]: copy paste :). Removed it.

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs (hunk 3 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.ResourceDetectors.Azure.Tests/AzureResourceDetectorTests.cs:1]: discussed offline. will keep it as it is for now. may separate it out later

File: src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs (hunk 7 above)
Comment (cijothomas) [reply src/OpenTelemetry.ResourceDetectors.Azure/AppServiceResourceDetector.cs:55]: we haven't written down any recommendation. null works. Empty also works. Empty looks a bit neater.

File: test/OpenTelemetry.ResourceDetectors.Azure.Tests/OpenTelemetry.ResourceDetectors.Azure.Tests.csproj [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>net7.0;net6.0;</TargetFrameworks>
```
Comment (cijothomas) [inline test/OpenTelemetry.ResourceDetectors.Azure.Tests/OpenTelemetry.ResourceDetectors.Azure.Tests.csproj:5]: net462

**Review by cijothomas — APPROVED** [review]

**Review by rajkumar-rangaraj — APPROVED** [review]

**Comment by julealgon** [comment]
@cijothomas this is a very interesting/useful package... any estimate for when we'll start to have RC/final versions of it? The implementation looks so simple, so I was surprised it was released as alpha still.

**Pull request merged**